{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5cba67e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from modlamp.descriptors import PeptideDescriptor, GlobalDescriptor\n",
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "import os, re, math, platform\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "from Bio import Seq, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7baa440d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<ipython-input-5-bdf21ded6e7a>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  if n is 1:\n",
      "<ipython-input-5-bdf21ded6e7a>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 2:\n",
      "<ipython-input-5-bdf21ded6e7a>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 3:\n"
     ]
    }
   ],
   "source": [
    "# Miscellaneous\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H',\n",
    "             'I', 'K', 'L', 'M', 'N', 'P', 'Q',\n",
    "             'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "n_gram statistics\n",
    "\"\"\"\n",
    "\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H', \n",
    "            'I', 'K', 'L', 'M', 'N', 'P', 'Q', \n",
    "            'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "def get_aan_corpus(n=2):\n",
    "    '''\n",
    "    Get AA corpus of n_gram (e.g. Di, Tri, etc.)\n",
    "    Output: AA n_gram corpus ((e.g. Di:400, Tri:3000, etc.))\n",
    "    '''\n",
    "    n_corpus = []\n",
    "    if n <= 2:\n",
    "        for i in _AALetter:\n",
    "            for j in _AALetter:\n",
    "               n_corpus.append(\"{}{}\".format(i, j))\n",
    "        return n_corpus\n",
    "    for i in get_aan_corpus(n - 1):\n",
    "        for j in _AALetter:\n",
    "            n_corpus.append(\"{}{}\".format(i, j))\n",
    "    return n_corpus\n",
    "\n",
    "\n",
    "def get_ngram_counts(seq, n=2):\n",
    "    '''\n",
    "    Get n_gram statistics\n",
    "    Input: peptide sequence and n\n",
    "    Ouput: n_gram statistic (dictionary) {A.A Corp: Counts}\n",
    "    '''\n",
    "    # Get the name of ngram feature\n",
    "    if n == 2:\n",
    "        prefix = 'DPC'\n",
    "    elif n == 3:\n",
    "        prefix = 'TPC'\n",
    "    else:\n",
    "        prefix = '{}gram'.format(n)\n",
    "\n",
    "    ngrams = [seq[i: i + n] for i in range(len(seq) - n + 1)]\n",
    "    n_corpus = get_aan_corpus(n)\n",
    "    ngram_stat = {}\n",
    "    for aa_ng in n_corpus:\n",
    "        ngram_stat['{}|{}'.format(prefix, aa_ng)] = ngrams.count(aa_ng) / len(ngrams) * 100\n",
    "    return ngram_stat\n",
    "\n",
    "\n",
    "def minSequenceLength(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(i[1]):\n",
    "            minLen = len(i[1])\n",
    "    return minLen\n",
    "\n",
    "\n",
    "def minSequenceLengthWithNormalAA(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(re.sub('-', '', i[1])):\n",
    "            minLen = len(re.sub('-', '', i[1]))\n",
    "    return minLen\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    input.fasta:      the input protein sequence file in fasta format.\n",
    "    k_space:          the gap of two amino acids, integer, defaule: 5\n",
    "    output:           the encoding file, default: 'encodings.tsv'\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def generateGroupPairs(groupKey):\n",
    "    gPair = {}\n",
    "    for key1 in groupKey:\n",
    "        for key2 in groupKey:\n",
    "            gPair['CKSAAGP|'+key1+'.'+key2] = 0\n",
    "    return gPair\n",
    "\n",
    "\n",
    "def cksaagp(fastas, gap = 5, **kw):\n",
    "    if gap < 0:\n",
    "        print('Error: the gap should be equal or greater than zero' + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    if minSequenceLength(fastas) < gap+2:\n",
    "        print('Error: all the sequence length should be greater than the (gap value) + 2 = ' + str(gap+2) + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    group = {'aliphatic': 'GAVLMI',\n",
    "             'aromatic': 'FYW',\n",
    "             'postivecharge': 'KRH',\n",
    "             'negativecharge': 'DE',\n",
    "             'uncharge': 'STCPNQ'}\n",
    "\n",
    "    AA = 'ARNDCQEGHILKMFPSTWYV'\n",
    "\n",
    "    groupKey = group.keys()\n",
    "\n",
    "    index = {}\n",
    "    for key in groupKey:\n",
    "        for aa in group[key]:\n",
    "            index[aa] = key\n",
    "\n",
    "    gPairIndex = []\n",
    "    for key1 in groupKey:\n",
    "        for key2 in groupKey:\n",
    "            gPairIndex.append('CKSAAGP|'+key1+'.'+key2)\n",
    "\n",
    "    encodings = []\n",
    "    header = ['#']\n",
    "    for g in range(gap + 1):\n",
    "        for p in gPairIndex:\n",
    "            header.append(p+'.gap'+str(g))\n",
    "    encodings.append(header)\n",
    "\n",
    "    for i in fastas:\n",
    "        name, sequence = i[0], re.sub('-', '', i[1])\n",
    "        code = [name]\n",
    "        for g in range(gap + 1):\n",
    "            gPair = generateGroupPairs(groupKey)\n",
    "            sum = 0\n",
    "            for p1 in range(len(sequence)):\n",
    "                p2 = p1 + g + 1\n",
    "                if p2 < len(sequence) and sequence[p1] in AA and sequence[p2] in AA:\n",
    "                    gPair['CKSAAGP|'+index[sequence[p1]]+'.'+index[sequence[p2]]] = gPair['CKSAAGP|'+index[sequence[p1]]+'.'+index[sequence[p2]]] + 1\n",
    "                    sum = sum + 1\n",
    "\n",
    "            if sum == 0:\n",
    "                for gp in gPairIndex:\n",
    "                    code.append(0)\n",
    "            else:\n",
    "                for gp in gPairIndex:\n",
    "                    code.append(gPair[gp] / sum)\n",
    "\n",
    "        encodings.append(code)\n",
    "\n",
    "    return encodings\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    input.fasta:      the input protein sequence file in fasta format.\n",
    "    lambda:           the lambda value, integer, defaule: 30\n",
    "    output:           the encoding file, default: 'encodings.tsv'\n",
    "\"\"\"\n",
    "\n",
    "def Rvalue(aa1, aa2, AADict, Matrix):\n",
    "    return sum([(Matrix[i][AADict[aa1]] - Matrix[i][AADict[aa2]]) ** 2 for i in range(len(Matrix))]) / len(Matrix)\n",
    "\n",
    "\n",
    "def paac(fastas, lambdaValue=30, w=0.05, **kw):\n",
    "    if minSequenceLengthWithNormalAA(fastas) < lambdaValue + 1:\n",
    "        print('Error: all the sequence length should be larger than the lambdaValue+1: ' + str(lambdaValue + 1) + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    dataFile = re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + r'\\data\\PAAC.txt' if platform.system() == 'Windows' else re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + '/data/PAAC.txt'\n",
    "    with open(dataFile) as f:\n",
    "        records = f.readlines()\n",
    "    AA = ''.join(records[0].rstrip().split()[1:])\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    AAProperty = []\n",
    "    AAPropertyNames = []\n",
    "    for i in range(1, len(records)):\n",
    "        array = records[i].rstrip().split() if records[i].rstrip() != '' else None\n",
    "        AAProperty.append([float(j) for j in array[1:]])\n",
    "        AAPropertyNames.append(array[0])\n",
    "\n",
    "    AAProperty1 = []\n",
    "    for i in AAProperty:\n",
    "        meanI = sum(i) / 20\n",
    "        fenmu = math.sqrt(sum([(j-meanI)**2 for j in i])/20)\n",
    "        AAProperty1.append([(j-meanI)/fenmu for j in i])\n",
    "\n",
    "    encodings = []\n",
    "    header = ['#']\n",
    "    for aa in AA:\n",
    "        header.append('PAAC|' + aa)\n",
    "    for n in range(1, lambdaValue + 1):\n",
    "        header.append('PAAC|lambda' + str(n))\n",
    "    encodings.append(header)\n",
    "\n",
    "    for i in fastas:\n",
    "        name, sequence = i[0], re.sub('-', '', i[1])\n",
    "        code = [name]\n",
    "        theta = []\n",
    "        for n in range(1, lambdaValue + 1):\n",
    "            theta.append(\n",
    "                sum([Rvalue(sequence[j], sequence[j + n], AADict, AAProperty1) for j in range(len(sequence) - n)]) / (\n",
    "                len(sequence) - n))\n",
    "        myDict = {}\n",
    "        for aa in AA:\n",
    "            myDict[aa] = sequence.count(aa)\n",
    "        code = code + [myDict[aa] / (1 + w * sum(theta)) for aa in AA]\n",
    "        code = code + [(w * j) / (1 + w * sum(theta)) for j in theta]\n",
    "        encodings.append(code)\n",
    "    return encodings\n",
    "\n",
    "\n",
    "def GAAC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphatic': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharge': 'KRH',\n",
    "\t\t'negativecharge': 'DE',\n",
    "\t\t'uncharge': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#']\n",
    "\tfor key in groupKey:\n",
    "\t\theader.append(\"GAAC|\"+key)\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\t\tcode = [name]\n",
    "\t\tcount = Counter(sequence)\n",
    "\t\tmyDict = {}\n",
    "\t\tfor key in groupKey:\n",
    "\t\t\tfor aa in group[key]:\n",
    "\t\t\t\tmyDict[key] = myDict.get(key, 0) + count[aa]\n",
    "\n",
    "\t\tfor key in groupKey:\n",
    "\t\t\tcode.append(myDict[key]/len(sequence))\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "\n",
    "def GDPC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphaticr': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharger': 'KRH',\n",
    "\t\t'negativecharger': 'DE',\n",
    "\t\t'uncharger': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\tbaseNum = len(groupKey)\n",
    "\tdipeptide = [g1+'.'+g2 for g1 in groupKey for g2 in groupKey]\n",
    "\n",
    "\tindex = {}\n",
    "\tfor key in groupKey:\n",
    "\t\tfor aa in group[key]:\n",
    "\t\t\tindex[aa] = key\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#'] + ['GDPC|'+dipname for dipname in dipeptide]\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\n",
    "\t\tcode = [name]\n",
    "\t\tmyDict = {}\n",
    "\t\tfor t in dipeptide:\n",
    "\t\t\tmyDict[t] = 0\n",
    "\n",
    "\t\tsum = 0\n",
    "\t\tfor j in range(len(sequence) - 2 + 1):\n",
    "\t\t\tmyDict[index[sequence[j]]+'.'+index[sequence[j+1]]] = myDict[index[sequence[j]]+'.'+index[sequence[j+1]]] + 1\n",
    "\t\t\tsum = sum +1\n",
    "\n",
    "\t\tif sum == 0:\n",
    "\t\t\tfor t in dipeptide:\n",
    "\t\t\t\tcode.append(0)\n",
    "\t\telse:\n",
    "\t\t\tfor t in dipeptide:\n",
    "\t\t\t\tcode.append(myDict[t]/sum)\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "\n",
    "def GTPC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphaticr': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharger': 'KRH',\n",
    "\t\t'negativecharger': 'DE',\n",
    "\t\t'uncharger': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\tbaseNum = len(groupKey)\n",
    "\ttriple = [g1+'.'+g2+'.'+g3 for g1 in groupKey for g2 in groupKey for g3 in groupKey]\n",
    "\n",
    "\tindex = {}\n",
    "\tfor key in groupKey:\n",
    "\t\tfor aa in group[key]:\n",
    "\t\t\tindex[aa] = key\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#'] + ['GTPC|'+tname for tname in triple]\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\n",
    "\t\tcode = [name]\n",
    "\t\tmyDict = {}\n",
    "\t\tfor t in triple:\n",
    "\t\t\tmyDict[t] = 0\n",
    "\n",
    "\t\tsum = 0\n",
    "\t\tfor j in range(len(sequence) - 3 + 1):\n",
    "\t\t\tmyDict[index[sequence[j]]+'.'+index[sequence[j+1]]+'.'+index[sequence[j+2]]] = myDict[index[sequence[j]]+'.'+index[sequence[j+1]]+'.'+index[sequence[j+2]]] + 1\n",
    "\t\t\tsum = sum +1\n",
    "\n",
    "\t\tif sum == 0:\n",
    "\t\t\tfor t in triple:\n",
    "\t\t\t\tcode.append(0)\n",
    "\t\telse:\n",
    "\t\t\tfor t in triple:\n",
    "\t\t\t\tcode.append(myDict[t]/sum)\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "'''\n",
    "Insert Iso_electric Point and net_charge(neutral) feature to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., iep, net_charge}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_phycs(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    #  Function for compute Isoelectric Point or net_charge of peptide\n",
    "    def get_ieq_nc(seq, is_iep=True):\n",
    "        protparam = PA(seq)\n",
    "        return protparam.isoelectric_point() if is_iep else protparam.charge_at_pH(7.0)\n",
    "\n",
    "    # Calculating IsoElectricPoints and NeutralCharge\n",
    "    data_size = seq_df.size\n",
    "    seq_df['PHYC|IEP'] = list(map(get_ieq_nc, seq_df['Sequence'], [True] * data_size))  # IsoElectricPoints\n",
    "    seq_df['PHYC|Net Charge'] = list(map(get_ieq_nc, seq_df['Sequence'], [False] * data_size))  # Charge(Neutral)\n",
    "\n",
    "    # Calculating hydrophobic moment (My assume all peptides are alpha-helix)\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'eisenberg')\n",
    "    descrpt.calculate_moment(window=1000, angle=100, modality='max')\n",
    "    seq_df['PHYC|Hydrophobic Moment'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating \"Hopp-Woods\" hydrophobicity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'hopp-woods')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Hydrophobicity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Energy of Transmembrane Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'tm_tend')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Transmembrane Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aromaticity\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aromaticity()\n",
    "    seq_df['PHYC|Aromacity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Levitt_alpha_helical Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'levitt_alpha')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Alpha Helical Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aliphatic Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aliphatic_index()\n",
    "    seq_df['PHYC|Aliphatic Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Boman Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.boman_index()\n",
    "    seq_df['PHYC|Boman Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "'''\n",
    "Insert Amino acid composition to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., AAC_Ax ... AAC_Yx}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_aac(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    # Compute AAC for peptide in specific A.A\n",
    "    def get_aac(seq, aa):\n",
    "        return seq.count(aa) / len(seq) * 100\n",
    "\n",
    "    # processing data_frame\n",
    "    data_size = seq_df.size\n",
    "    for ll in _AALetter:\n",
    "        seq_df['AAC|{}'.format(ll)] = list(map(get_aac, seq_df['Sequence'], [ll] * data_size))\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "'''\n",
    "Insert n_grams Descriptor to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., ngram_(1), .., ngram(20 ** n)}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_ngrams(seq_df, n=2):\n",
    "    seq_df = seq_df.copy()\n",
    "    data_size = seq_df.size\n",
    "\n",
    "    ngrams_df = list(map(get_ngram_counts, seq_df['Sequence'], [n] * data_size))\n",
    "    ngrams_df = pd.DataFrame(ngrams_df)  # Convert ngrams features to pd.DataFrame\n",
    "    seq_df = pd.concat([seq_df, ngrams_df], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert CKSAAGP Descriptor to the sequence data_frame\n",
    "(Composition of k-spaced amino acid pairs)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_cksaagp(seq_df, gap=2):\n",
    "    seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    encoding = cksaagp(fastas, gap=gap)\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert PAAC Descriptor to the sequence data_frame\n",
    "(Pseudo Amino Acid Composition)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_paac(seq_df, lamb=3, w=0.1):\n",
    "    seq_df = seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    encoding = paac(fastas, lambdaValue=lamb, w=w)\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert Grouped n-gram Descriptor to the sequence data_frame\n",
    "(Pseudo Amino Acid Composition)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_gngram(seq_df, n=1):\n",
    "    seq_df = seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    # encoding = paac(fastas, lambdaValue=lamb, w=w)\n",
    "    if n is 1:\n",
    "        encoding = GAAC(fastas)\n",
    "    elif n is 2:\n",
    "        encoding = GDPC(fastas)\n",
    "    elif n is 3:\n",
    "        encoding = GTPC(fastas)\n",
    "    else:\n",
    "        raise Warning(\"Invalid n-grams, no features added\")\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "def construct_features(seq_df, paaclamb=4, paacw=0.5):\n",
    "    \"\"\"\n",
    "    Construct Features for the AVPIden. We first investigated physiochemical feautres, AAC features, DiC features, \n",
    "    CKSAAGP features, PAAC features, and PHYC features.\n",
    "    Parameters are pre-set according to the sequence identities.\n",
    "    \"\"\"\n",
    "    seq_df = insert_aac(seq_df)\n",
    "    seq_df = insert_ngrams(seq_df, n=2)\n",
    "    seq_df = insert_cksaagp(seq_df, gap=3) # As the maximum motif length = 5.\n",
    "    seq_df = insert_paac(seq_df, lamb=paaclamb, w=paacw)\n",
    "    seq_df = insert_phycs(seq_df)\n",
    "\n",
    "    return seq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6569f773",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta(df, file_path, abbr_columns=None):\n",
    "    \"\"\"\n",
    "    Save dataframe to a .fasta file, the df should contain at least columns named \"Id\" and \"Sequence\"\n",
    "    \n",
    "    df: dataframe for saving .fasta\n",
    "    file_path: path(string) for saving the fasta file\n",
    "    abbr_columns: string columns for adding abbreviations. Multiple abbr are splited by '|'.\n",
    "    \"\"\"\n",
    "    Seqrecords = [SeqIO.SeqRecord(id=row['Id'], \n",
    "                              seq=Seq.Seq(row['Sequence']), \n",
    "                              description='|'.join(row[abbr_columns] if abbr_columns is not None else \"\")) \\\n",
    "             for idn, row in df.iterrows()]\n",
    "    with open(file_path, 'w+') as fhandle:\n",
    "        SeqIO.write(Seqrecords, fhandle, \"fasta-2line\")\n",
    "        print(\"Saved {:d} sequences.\".format(len(Seqrecords)))\n",
    "\n",
    "\n",
    "def read_fasta(fname):\n",
    "    '''\n",
    "    Read fasta file to dictionary\n",
    "    Input: path name of fasta\n",
    "    Output: dataframe of Peptide Seq {ID1: Seq1, ID2: Seq2,...}\n",
    "    '''\n",
    "    with open(fname, \"rU\") as f:\n",
    "        seq_dict = [(record.id, record.seq._data) for record in SeqIO.parse(f, \"fasta\")]\n",
    "    seq_df = pd.DataFrame(data=seq_dict, columns=[\"Id\", \"Sequence\"])\n",
    "    return seq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf2566c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-6-e2b83a1f0503>:24: DeprecationWarning: 'U' mode is deprecated\n",
      "  with open(fname, \"rU\") as f:\n"
     ]
    }
   ],
   "source": [
    "seq_fasta = read_fasta(\"ATP_sequence.fasta\")\n",
    "seq_feature = construct_features(seq_fasta)\n",
    "seq_feature.to_csv(\"ATP_feature.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
